### PR TITLE
feat(trie): sparse trie methods for trie task integration

### PR DESCRIPTION
**Pull request by shekhirin**
Extracted from https://github.com/paradigmxyz/reth/pull/12676 + a small cleanup after https://github.com/paradigmxyz/reth/pull/12663/files

### CODE REVIEW

File: crates/trie/sparse/src/state.rs [hunk 1]
```diff
@@ -124,11 +124,38 @@ impl SparseStateTrie {
         Ok(())
     }
 
+    /// Remove the leaf node.
+    pub fn remove_leaf(&mut self, path: &Nibbles) -> SparseStateTrieResult<()> {
```
Comment (rkrasiuk) [inline crates/trie/sparse/src/state.rs:128]: mb? 
```suggestion
    pub fn remove_account_leaf(&mut self, path: &Nibbles) -> SparseStateTrieResult<()> {
```

File: crates/trie/sparse/src/state.rs [hunk 2]
```diff
@@ -124,11 +124,38 @@ impl SparseStateTrie {
...
+        slot: Nibbles,
+        value: Vec<u8>,
+    ) -> SparseStateTrieResult<()> {
+        self.storages.entry(address).or_default().update_leaf(slot, value)?;
+        Ok(())
+    }
+
+    /// Wipe the storage trie at the provided address.
+    pub fn wipe_storage(&mut self, address: B256) {
+        self.storages.remove(&address);
```
Comment (rkrasiuk) [inline crates/trie/sparse/src/state.rs:156]: is this correct? we still need to preserve the fact that the storage was wiped and wipe it from the database on commit

File: crates/trie/sparse/src/state.rs (hunk 1 above)
Comment (shekhirin) [reply crates/trie/sparse/src/state.rs]: agreed

File: crates/trie/sparse/src/state.rs (hunk 2 above)
Comment (shekhirin) [reply crates/trie/sparse/src/state.rs]: nvm it's still incorrect, turning into a draft for now

File: crates/trie/sparse/src/trie.rs [hunk 3]
```diff
@@ -553,6 +572,17 @@ impl RevealedSparseTrie {
...
+    /// Wipe the trie, removing all values and nodes, and replacing the root with an empty node.
+    pub fn wipe(&mut self) -> SparseTrieResult<()> {
+        let paths = self.values.drain().map(|(k, _)| k).collect::<Vec<_>>();
+        // TODO: can we optimize the batch deletion here?
+        for path in paths {
+            self.remove_leaf(&path)?;
+        }
+        debug_assert_eq!(self.nodes, HashMap::from_iter([(Nibbles::default(), SparseNode::Empty)]));
+        Ok(())
+    }
```
Comment (rkrasiuk) [inline crates/trie/sparse/src/trie.rs:584]: mb?
```suggestion
    /// Wipe the trie, removing all values and nodes, and replacing the root with an empty node.
    pub fn wipe(&mut self) {
        *self = Self::default();
    }
```

Comment (shekhirin) [reply crates/trie/sparse/src/trie.rs:584]: yeah i did it that way at first, but the branch node retaining logic in https://github.com/paradigmxyz/reth/pull/12291 relies on calling the `remove_leaf`. That's why I added this TODO https://github.com/paradigmxyz/reth/blob/e5e052d75dd34ef8dd26999adabbc32f0439aafd/crates/trie/sparse/src/trie.rs#L578

Comment (rkrasiuk) [reply crates/trie/sparse/src/trie.rs:611]: @shekhirin but we should just wipe any intermediate node in the database, it's fine if this diverges from the HB behavior

Comment (shekhirin) [reply crates/trie/sparse/src/trie.rs]: we still need to retain the removed branch nodes, I adjusted your code a bit, PTAL

Comment (shekhirin) [reply crates/trie/sparse/src/trie.rs]: nvm, there's a separate flag in updates struct that we can use, so your approach works fine https://github.com/paradigmxyz/reth/blob/7f5fd80cb7d69d1e5d14abf6851111adae762970/crates/trie/db/src/trie_cursor.rs#L126-L129

File: crates/trie/sparse/src/trie.rs [hunk 4]
```diff
@@ -580,6 +599,13 @@ impl RevealedSparseTrie {
         Ok(nodes)
     }
 
+    /// Wipe the trie, removing all values and nodes, and replacing the root with an empty node.
+    pub fn wipe(&mut self) -> SparseTrieResult<()> {
+        *self = Self::default();
+        self.prefix_set = PrefixSetMut::all();
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/trie/sparse/src/trie.rs]: fixed

**Review by rkrasiuk — APPROVED** [review]

**Pull request merged**